File: src/angel/clients/github.py
import time
from collections.abc import Callable
from pathlib import Path
from typing import Any, NamedTuple, Protocol

import httpx
import jwt

from angel.env import load_env_var
from angel.logs import get_logger

log = get_logger(__name__)


def _get_all(
    http_client: httpx.Client,
    url: str,
    *,
    params: dict[str, Any] | None = None,
    headers: dict[str, str] | None = None,
    key: str | None = None,
) -> list[Any]:
    items: list[Any] = []
    while True:
        response = http_client.get(url, params=params, headers=headers)
        response.raise_for_status()
        items.extend(response.json()[key] if key else response.json())
        next_link = response.links.get("next")
        if not next_link:
            return items
        url = next_link["url"]
        params = None


class PullRequest(NamedTuple):
    repository: str
    number: int
    title: str
    body: str
    head_sha: str


class Issue(NamedTuple):
    repository: str
    number: int
    title: str
    body: str
    updated_at: str


class Comment(NamedTuple):
    author: str
    body: str
    created_at: str


class GitHubClient(Protocol):
    def list_repositories(self) -> list[str]: ...
    def ensure_label(self, repository: str, label: str, color: str) -> None: ...
    def list_open_pull_requests(
        self, repository: str, label: str
    ) -> list[PullRequest]: ...
    def fetch_diff(self, pull_request: PullRequest) -> str: ...
    def list_open_issues(self, repository: str, label: str) -> list[Issue]: ...
    def list_comments(self, target: PullRequest | Issue) -> list[Comment]: ...
    def post_comment(self, target: PullRequest | Issue, body: str) -> None: ...


class TokenProvider(Protocol):
    def __call__(self, repository: str) -> str: ...
    def list_repositories(self) -> list[str]: ...


class MissingGitHubCredentialsError(RuntimeError):
    pass


GITHUB_BASE_URL = "https://api.github.com"


class InstallationTokenProvider:
    def __init__(
        self,
        http_client: httpx.Client,
        app_id: str,
        private_key: str,
        now: Callable[[], int] = lambda: int(time.time()),
    ) -> None:
        self.http_client = http_client
        self.app_id = app_id
        self.private_key = private_key
        self.now = now
        self.tokens: dict[str, str] = {}

    def _jwt_header(self) -> dict[str, str]:
        now = self.now()
        jwt_token = jwt.encode(
            {"iat": now - 60, "exp": now + 540, "iss": self.app_id},
            self.private_key,
            algorithm="RS256",
        )
        return {"Authorization": f"Bearer {jwt_token}"}

    def _create_token(self, installation_id: int, headers: dict[str, str]) -> str:
        response = self.http_client.post(
            f"/app/installations/{installation_id}/access_tokens", headers=headers
        )
        response.raise_for_status()
        return response.json()["token"]

    def __call__(self, repository: str) -> str:
        if repository in self.tokens:
            return self.tokens[repository]

        headers = self._jwt_header()

        installation_response = self.http_client.get(
            f"/repos/{repository}/installation", headers=headers
        )
        installation_response.raise_for_status()
        installation_id = installation_response.json()["id"]

        token = self._create_token(installation_id, headers)
        self.tokens[repository] = token
        return token

    def list_repositories(self) -> list[str]:
        headers = self._jwt_header()
        installations = _get_all(
            self.http_client,
            "/app/installations",
            params={"per_page": 100},
            headers=headers,
        )

        repositories = []
        for installation in installations:
            token = self._create_token(installation["id"], headers)
            for repository in _get_all(
                self.http_client,
                "/installation/repositories",
                params={"per_page": 100},
                headers={"Authorization": f"Bearer {token}"},
                key="repositories",
            ):
                self.tokens[repository["full_name"]] = token
                repositories.append(repository["full_name"])
        return repositories


class HttpGitHubClient:
    def __init__(
        self, http_client: httpx.Client, token_provider: TokenProvider
    ) -> None:
        self.http_client = http_client
        self.token_provider = token_provider

    def _auth_header(self, repository: str) -> dict[str, str]:
        return {"Authorization": f"Bearer {self.token_provider(repository)}"}

    def list_repositories(self) -> list[str]:
        return self.token_provider.list_repositories()

    def ensure_label(self, repository: str, label: str, color: str) -> None:
        response = self.http_client.get(
            f"/repos/{repository}/labels/{label}",
            headers=self._auth_header(repository),
        )
        if response.status_code == 404:
            create_response = self.http_client.post(
                f"/repos/{repository}/labels",
                json={"name": label, "color": color},
                headers=self._auth_header(repository),
            )
            create_response.raise_for_status()
            return
        response.raise_for_status()

    def list_open_pull_requests(self, repository: str, label: str) -> list[PullRequest]:
        items = _get_all(
            self.http_client,
            f"/repos/{repository}/pulls",
            params={"state": "open", "per_page": 100},
            headers=self._auth_header(repository),
        )
        return [
            PullRequest(
                repository,
                item["number"],
                item["title"],
                item["body"] or "",
                item["head"]["sha"],
            )
            for item in items
            if label in {item_label["name"] for item_label in item["labels"]}
        ]

    def fetch_diff(self, pull_request: PullRequest) -> str:
        response = self.http_client.get(
            f"/repos/{pull_request.repository}/pulls/{pull_request.number}",
            headers={
                "Accept": "application/vnd.github.diff",
                **self._auth_header(pull_request.repository),
            },
        )
        response.raise_for_status()
        return response.text

    def list_open_issues(self, repository: str, label: str) -> list[Issue]:
        items = _get_all(
            self.http_client,
            f"/repos/{repository}/issues",
            params={"state": "open", "per_page": 100, "labels": label},
            headers=self._auth_header(repository),
        )
        return [
            Issue(
                repository,
                item["number"],
                item["title"],
                item["body"] or "",
                item["updated_at"],
            )
            for item in items
            if "pull_request" not in item
        ]

    def list_comments(self, target: PullRequest | Issue) -> list[Comment]:
        items = _get_all(
            self.http_client,
            f"/repos/{target.repository}/issues/{target.number}/comments",
            params={"per_page": 100},
            headers=self._auth_header(target.repository),
        )
        return [
            Comment(item["user"]["login"], item["body"] or "", item["created_at"])
            for item in items
        ]

    def post_comment(self, target: PullRequest | Issue, body: str) -> None:
        response = self.http_client.post(
            f"/repos/{target.repository}/issues/{target.number}/comments",
            json={"body": body},
            headers=self._auth_header(target.repository),
        )
        response.raise_for_status()


class DryRunGitHubClient:
    def __init__(self, inner: GitHubClient) -> None:
        self.inner = inner

    def list_repositories(self) -> list[str]:
        return self.inner.list_repositories()

    def ensure_label(self, repository: str, label: str, color: str) -> None:
        log.info(
            "dry run: skipping ensure_label",
            dry_run=True,
            repository=repository,
            label=label,
        )

    def list_open_pull_requests(self, repository: str, label: str) -> list[PullRequest]:
        return self.inner.list_open_pull_requests(repository, label)

    def fetch_diff(self, pull_request: PullRequest) -> str:
        return self.inner.fetch_diff(pull_request)

    def list_open_issues(self, repository: str, label: str) -> list[Issue]:
        return self.inner.list_open_issues(repository, label)

    def list_comments(self, target: PullRequest | Issue) -> list[Comment]:
        return self.inner.list_comments(target)

    def post_comment(self, target: PullRequest | Issue, body: str) -> None:
        log.info(
            "dry run: skipping post_comment",
            dry_run=True,
            repository=target.repository,
            number=target.number,
            body_length=len(body),
        )


def build_github_client() -> GitHubClient:
    app_id = load_env_var("ANGEL_GITHUB_APP_ID")
    private_key_path = load_env_var("ANGEL_GITHUB_PRIVATE_KEY_PATH")
    if not app_id or not private_key_path:
        missing = [
            name
            for name, value in (
                ("ANGEL_GITHUB_APP_ID", app_id),
                ("ANGEL_GITHUB_PRIVATE_KEY_PATH", private_key_path),
            )
            if not value
        ]
        raise MissingGitHubCredentialsError(
            " ".join(f"{name} environment variable is not set." for name in missing)
        )

    private_key = Path(private_key_path).expanduser().read_text()

    http_client = httpx.Client(
        base_url=GITHUB_BASE_URL,
        headers={"X-GitHub-Api-Version": "2022-11-28"},
    )
    return HttpGitHubClient(
        http_client, InstallationTokenProvider(http_client, app_id, private_key)
    )

File: apps/server/clients/github.ts
import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import jwt from "jsonwebtoken";
import { loadEnvVar } from "../../../packages/shared/env.js";
import { getLogger } from "../../../packages/shared/logs.js";
import { type FetchLike, getAll, HttpClient } from "./http.js";

const log = getLogger("angel.clients.github");

export const GITHUB_BASE_URL = "https://api.github.com";
export const GITHUB_API_VERSION = "2022-11-28";

const PER_PAGE = 100;
const JWT_LEEWAY_SECONDS = 60;
const JWT_LIFETIME_SECONDS = 540;

export interface PullRequest {
  readonly repository: string;
  readonly number: number;
  readonly title: string;
  readonly body: string;
  readonly headSha: string;
}

export interface Issue {
  readonly repository: string;
  readonly number: number;
  readonly title: string;
  readonly body: string;
  readonly updatedAt: string;
}

export interface Comment {
  readonly author: string;
  readonly body: string;
  readonly createdAt: string;
}

export type ReviewTarget = PullRequest | Issue;

export interface GitHubClient {
  listRepositories(): Promise<string[]>;
  ensureLabel(repository: string, label: string, color: string): Promise<void>;
  listOpenPullRequests(repository: string, label: string): Promise<PullRequest[]>;
  fetchDiff(pullRequest: PullRequest): Promise<string>;
  listOpenIssues(repository: string, label: string): Promise<Issue[]>;
  listComments(target: ReviewTarget): Promise<Comment[]>;
  postComment(target: ReviewTarget, body: string): Promise<void>;
}

export interface TokenProvider {
  tokenFor(repository: string): Promise<string>;
  listRepositories(): Promise<string[]>;
}

export class MissingGitHubCredentialsError extends Error {
  override readonly name = "MissingGitHubCredentialsError";
}

interface ApiLabelRef {
  readonly name: string;
}

interface ApiPullRequest {
  readonly number: number;
  readonly title: string;
  readonly body: string | null;
  readonly head: { readonly sha: string };
  readonly labels: readonly ApiLabelRef[];
}

interface ApiIssue {
  readonly number: number;
  readonly title: string;
  readonly body: string | null;
  readonly updated_at: string;
  readonly pull_request?: unknown;
}

interface ApiComment {
  readonly user: { readonly login: string };
  readonly body: string | null;
  readonly created_at: string;
}

interface ApiInstallation {
  readonly id: number;
}

interface ApiRepository {
  readonly full_name: string;
}

interface ApiAccessToken {
  readonly token: string;
}

export class InstallationTokenProvider implements TokenProvider {
  private readonly tokens = new Map<string, string>();

  constructor(
    private readonly httpClient: HttpClient,
    private readonly appId: string,
    private readonly privateKey: string,
    private readonly now: () => number = () => Math.floor(Date.now() / 1000),
  ) {}

  private jwtHeader(): Record<string, string> {
    const now = this.now();
    const token = jwt.sign(
      {
        iat: now - JWT_LEEWAY_SECONDS,
        exp: now + JWT_LIFETIME_SECONDS,
        iss: this.appId,
      },
      this.privateKey,
      { algorithm: "RS256" },
    );
    return { authorization: `Bearer ${token}` };
  }

  private async createToken(
    installationId: number,
    headers: Record<string, string>,
  ): Promise<string> {
    const response = (
      await this.httpClient.post(`/app/installations/${installationId}/access_tokens`, {
        headers,
      })
    ).ensureOk();
    return (await response.json<ApiAccessToken>()).token;
  }

  async tokenFor(repository: string): Promise<string> {
    const cached = this.tokens.get(repository);
    if (cached !== undefined) {
      return cached;
    }
    const headers = this.jwtHeader();
    const response = (
      await this.httpClient.get(`/repos/${repository}/installation`, { headers })
    ).ensureOk();
    const { id } = await response.json<ApiInstallation>();

    const token = await this.createToken(id, headers);
    this.tokens.set(repository, token);
    return token;
  }

  async listRepositories(): Promise<string[]> {
    const headers = this.jwtHeader();
    const installations = await getAll<ApiInstallation>(
      this.httpClient,
      "/app/installations",
      { params: { per_page: PER_PAGE }, headers },
    );

    const repositories: string[] = [];
    for (const installation of installations) {
      const token = await this.createToken(installation.id, headers);
      const owned = await getAll<ApiRepository>(
        this.httpClient,
        "/installation/repositories",
        {
          params: { per_page: PER_PAGE },
          headers: { authorization: `Bearer ${token}` },
          key: "repositories",
        },
      );
      for (const repository of owned) {
        this.tokens.set(repository.full_name, token);
        repositories.push(repository.full_name);
      }
    }
    return repositories;
  }
}

export class HttpGitHubClient implements GitHubClient {
  constructor(
    readonly httpClient: HttpClient,
    private readonly tokenProvider: TokenProvider,
  ) {}

  private async authHeader(repository: string): Promise<Record<string, string>> {
    return { authorization: `Bearer ${await this.tokenProvider.tokenFor(repository)}` };
  }

  listRepositories(): Promise<string[]> {
    return this.tokenProvider.listRepositories();
  }

  async ensureLabel(repository: string, label: string, color: string): Promise<void> {
    const response = await this.httpClient.get(`/repos/${repository}/labels/${label}`, {
      headers: await this.authHeader(repository),
    });
    if (response.status === 404) {
      (
        await this.httpClient.post(`/repos/${repository}/labels`, {
          json: { name: label, color },
          headers: await this.authHeader(repository),
        })
      ).ensureOk();
      return;
    }
    response.ensureOk();
  }

  async listOpenPullRequests(
    repository: string,
    label: string,
  ): Promise<PullRequest[]> {
    const items = await getAll<ApiPullRequest>(
      this.httpClient,
      `/repos/${repository}/pulls`,
      {
        params: { state: "open", per_page: PER_PAGE },
        headers: await this.authHeader(repository),
      },
    );
    return items
      .filter((item) => item.labels.some((each) => each.name === label))
      .map((item) => ({
        repository,
        number: item.number,
        title: item.title,
        body: item.body ?? "",
        headSha: item.head.sha,
      }));
  }

  async fetchDiff(pullRequest: PullRequest): Promise<string> {
    const response = (
      await this.httpClient.get(
        `/repos/${pullRequest.repository}/pulls/${pullRequest.number}`,
        {
          headers: {
            accept: "application/vnd.github.diff",
            ...(await this.authHeader(pullRequest.repository)),
          },
        },
      )
    ).ensureOk();
    return response.text();
  }

  async listOpenIssues(repository: string, label: string): Promise<Issue[]> {
    const items = await getAll<ApiIssue>(
      this.httpClient,
      `/repos/${repository}/issues`,
      {
        params: { state: "open", per_page: PER_PAGE, labels: label },
        headers: await this.authHeader(repository),
      },
    );
    return items
      .filter((item) => item.pull_request === undefined)
      .map((item) => ({
        repository,
        number: item.number,
        title: item.title,
        body: item.body ?? "",
        updatedAt: item.updated_at,
      }));
  }

  async listComments(target: ReviewTarget): Promise<Comment[]> {
    const items = await getAll<ApiComment>(
      this.httpClient,
      `/repos/${target.repository}/issues/${target.number}/comments`,
      {
        params: { per_page: PER_PAGE },
        headers: await this.authHeader(target.repository),
      },
    );
    return items.map((item) => ({
      author: item.user.login,
      body: item.body ?? "",
      createdAt: item.created_at,
    }));
  }

  async postComment(target: ReviewTarget, body: string): Promise<void> {
    (
      await this.httpClient.post(
        `/repos/${target.repository}/issues/${target.number}/comments`,
        { json: { body }, headers: await this.authHeader(target.repository) },
      )
    ).ensureOk();
  }
}

export class DryRunGitHubClient implements GitHubClient {
  constructor(private readonly inner: GitHubClient) {}

  listRepositories(): Promise<string[]> {
    return this.inner.listRepositories();
  }

  async ensureLabel(repository: string, label: string, _color: string): Promise<void> {
    log.info("dry run: skipping ensure_label", {
      dry_run: true,
      repository,
      label,
    });
  }

  listOpenPullRequests(repository: string, label: string): Promise<PullRequest[]> {
    return this.inner.listOpenPullRequests(repository, label);
  }

  fetchDiff(pullRequest: PullRequest): Promise<string> {
    return this.inner.fetchDiff(pullRequest);
  }

  listOpenIssues(repository: string, label: string): Promise<Issue[]> {
    return this.inner.listOpenIssues(repository, label);
  }

  listComments(target: ReviewTarget): Promise<Comment[]> {
    return this.inner.listComments(target);
  }

  async postComment(target: ReviewTarget, body: string): Promise<void> {
    log.info("dry run: skipping post_comment", {
      dry_run: true,
      repository: target.repository,
      number: target.number,
      body_length: body.length,
    });
  }
}

function expandUser(path: string): string {
  return path.startsWith("~/") ? join(homedir(), path.slice(2)) : path;
}

export function buildGithubClient(fetchImpl?: FetchLike): HttpGitHubClient {
  const appId = loadEnvVar("ANGEL_GITHUB_APP_ID");
  const privateKeyPath = loadEnvVar("ANGEL_GITHUB_PRIVATE_KEY_PATH");
  if (!appId || !privateKeyPath) {
    const missing = (
      [
        ["ANGEL_GITHUB_APP_ID", appId],
        ["ANGEL_GITHUB_PRIVATE_KEY_PATH", privateKeyPath],
      ] as const
    )
      .filter(([, value]) => !value)
      .map(([name]) => `${name} environment variable is not set.`);
    throw new MissingGitHubCredentialsError(missing.join(" "));
  }

  const privateKey = readFileSync(expandUser(privateKeyPath), "utf8");
  const httpClient = new HttpClient(
    GITHUB_BASE_URL,
    { "x-github-api-version": GITHUB_API_VERSION },
    fetchImpl,
  );
  return new HttpGitHubClient(
    httpClient,
    new InstallationTokenProvider(httpClient, appId, privateKey),
  );
}
